### PR TITLE
Add support for the Nordic nRF52840

### PR DESCRIPTION
**Pull request by toshiba-bril**
This is a port for the Nordic Semiconductor nRF52840 SoC developed for Toshiba's Bristol Research and Innovation Laboratory (BRIL).

The base support is based on the existing nrf52dk platform, with any references to the SoftDevice removed.
The radio driver depends only on the low-level HAL `nrf_radio` methods. 
Both TSCH and CSMA are supported. Interoperability has been tested with a TI CC2650 LaunchPad.

Support is provided for the PCA10056 nRF52840-DK and the PCA10059 nRF52840 dongle. 
No native USB support is provided, so debug prints from the dongle would have to be obtained from the UART pins.

The port uses the nRF5 SDK 16.0, a stripped version of which is provided as a submodule under `arch/cpu/nrf52840/lib/nrf52-sdk`.

The following make targets are provided:
* `.upload` - Upload to one connected nRF52840 board using nrfjprog
* `.upload-all` - Upload to all connected nRF52840 boards using nrfjprog
* `.dfu-upload` - Upload to one connected nRF52840 dongle using nrfutil

A wiki page for this is in the works.

### CODE REVIEW

File: arch/platform/nrf52840/dk/nrf52840-board-def.h [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+#define LEDS_CONF_RED       LEDS_3
+#define LEDS_CONF_BLUE      LEDS_4
+
+#define LEDS_CONF_ALL  (LEDS_1 | LEDS_2 | LEDS_3 | LEDS_4)
+
+/**
+ * \brief If set to 1 then LED1 and LED2 are used by the
+ *        platform to indicate BLE connection state.
+ */
+#define PLATFORM_INDICATE_BLE_STATE             1
```
Comment (g-oikonomou) [inline arch/platform/nrf52840/dk/nrf52840-board-def.h:81]: Do we need that?

File: arch/platform/nrf52840/dk/Makefile.dk [hunk 2]
```diff
@@ -0,0 +1,2 @@
+CONTIKI_TARGET_DIRS += dk dk/dev
```
Comment (g-oikonomou) [inline arch/platform/nrf52840/dk/Makefile.dk:1]: To be honest given how small the number of files here, I'd have dropped everything under `.` without the `dev` subdir

File: arch/platform/nrf52840/dk/dev/temperature-sensor.c [hunk 3]
```diff
@@ -0,0 +1,104 @@
+/*
```
Comment (g-oikonomou) [inline arch/platform/nrf52840/dk/dev/temperature-sensor.c:1]: We have that on the dongle too, right?

File: arch/platform/nrf52840/dongle/Makefile.dongle [hunk 4]
```diff
@@ -0,0 +1,2 @@
+CONTIKI_TARGET_DIRS += dongle dongle/dev
```
Comment (g-oikonomou) [inline arch/platform/nrf52840/dongle/Makefile.dongle:1]: ditto

File: arch/platform/nrf52840/Makefile.nrf52840 [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  $(error CONTIKI not defined! You must specify where CONTIKI resides!)
+endif
+
+### Board selection
+BOARD ?= dk
+BOARDS = dk dongle
+
+#CONTIKI_TARGET_DIRS += . dev config
+#CONTIKI_SOURCEFILES += platform.c nrf52840dk-sensors.c
+#CONTIKI_SOURCEFILES += button-sensor.c temperature-sensor.c
```
Comment (g-oikonomou) [inline arch/platform/nrf52840/Makefile.nrf52840:11]: Lose if not required

File: arch/cpu/nrf52840/Makefile.nrf52840 [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+else
+	$(MAKE) HEX_FILE=$< -j $(NUMPAR) nrf-upload-sequence
+endif
+
+# DFU Upload for dongles
+
+DFU_FLAGS = -p /dev/ttyACM0
+ifdef PORT
+  DFU_FLAGS = -p $(PORT)
+endif
```
Comment (g-oikonomou) [inline arch/cpu/nrf52840/Makefile.nrf52840:136]:
```
PORT ?= /dev/ttyACM0
DFU_FLAGS = -p $(PORT)
```
?

File: arch/cpu/nrf52840/Makefile.nrf52840 [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+ifeq ($(NRF52840_USE_RTT),1)
+### Suppress the existing debug I/O in os/lib
+CFLAGS += -DNRF52840_USE_RTT=1
+MAKE_WITH_LIB_DBG_IO = 0
+CONTIKI_TARGET_DIRS += rtt
+CONTIKI_CPU_SOURCEFILES += rtt-printf.c segger-rtt.c segger-rtt-printf.c
+else
+CONTIKI_CPU_SOURCEFILES += dbg.c
+endif
```
Comment (g-oikonomou) [inline arch/cpu/nrf52840/Makefile.nrf52840:17]: Can you indent here inside the conditionals?

File: arch/cpu/nrf52840/Makefile.nrf52840 [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+# Boards
+ifeq ($(BOARD), dongle)
+LDSCRIPT ?= $(CONTIKI_CPU)/ld/nrf52840-bootloader.ld
+CFLAGS += -DBOARD_PCA10059
+else
+LDSCRIPT ?= $(CONTIKI_CPU)/ld/nrf52840.ld
+CFLAGS += -DBOARD_PCA10056
+endif
```
Comment (g-oikonomou) [inline arch/cpu/nrf52840/Makefile.nrf52840:26]: Can you indent here inside the conditionals?

File: arch/cpu/nrf52840/lpm.h [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ */
+static inline void
+lpm_drop(void)
+{
+  int_master_status_t status;
+  int abort;
+  status = critical_enter();
+  abort = process_nevents();
+  if(!abort) {
```
Comment (g-oikonomou) [inline arch/cpu/nrf52840/lpm.h:60]: `ENERGEST_SWITCH(ENERGEST_TYPE_CPU, ENERGEST_TYPE_LPM);`

File: arch/cpu/nrf52840/lpm.h [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    __WFI();
```
Comment (g-oikonomou) [inline arch/cpu/nrf52840/lpm.h:61]: `ENERGEST_SWITCH(ENERGEST_TYPE_LPM, ENERGEST_TYPE_CPU);`

File: arch/cpu/nrf52840/lpm.h [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+  }
+  critical_exit(status);
+}
+#endif /* DEV_LPM_H_ */
```
Comment (g-oikonomou) [inline arch/cpu/nrf52840/lpm.h:65]: `LPM_H_`

**Comment by atiselsts** [comment]
Thanks for this!

A couple questions:
- How well tested this is? (TSCH, CSMA, how many nodes...)
- Any plans for BLE support in the future?

File: arch/cpu/nrf52840/Makefile.nrf52840 [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+# This approach is similar to the sky platform
+
+# Number of boards to program in parallel
+NUMPAR=4
+
+nrf-upload-snr.%:
+	nrfjprog -f nrf52 --snr $* --sectorerase --program $(HEX_FILE)
+	nrfjprog -f nrf52 --snr $* --reset
+
+nrf-upload-sequence: $(foreach SNR, $(shell nrfjprog -i), nrf-upload-snr.$(SNR))
```
Comment (g-oikonomou) [inline arch/cpu/nrf52840/Makefile.nrf52840:121]: I am not particularly keen on any of those `shell` things in this Makefile, but this rule is particularly problematic and causes the behaviour below.

```
~/contiki-ng/examples/hello-world (pr/1310)$ make TARGET=nrf52840 savetarget
make: nrfjprog: Command not found
saving Makefile.target
~/contiki-ng/examples/hello-world (pr/1310 *)$ make TARGET=nrf52840 boards
make: nrfjprog: Command not found
dk dongle (current: dk)
```

Personally I would simply do something like:
```
NRFJPROG ?= nrfjprog
NRFUTIL ?= nrfutil
```
and then remove the shell stuff.

This rule could then be (untested)
```
nrf-upload-sequence:
	$(foreach SNR, $(shell $(NRFJPROG) -i), nrf-upload-snr.$(SNR))
	@echo Done
```

File: arch/cpu/nrf52840/Makefile.nrf52840 [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+endif
+
+# DFU Upload for dongles
+
+DFU_FLAGS = -p /dev/ttyACM0
+ifdef PORT
+  DFU_FLAGS = -p $(PORT)
+endif
+
+%.dfu-upload: $(OUT_HEX) %.$(TARGET)
```
Comment (g-oikonomou) [inline arch/cpu/nrf52840/Makefile.nrf52840:138]: `%.dfu-upload: $(OUT_HEX)` should be enough here

File: arch/cpu/nrf52840/Makefile.nrf52840 [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+
+include $(ARCH_PATH)/cpu/arm/cortex-m/cm4/Makefile.cm4
+
+# Upload via J-Link using nrfjprog 
+
+ifdef NRF52_UPLOAD_SN
+	NRFJPROG_FLAGS += --snr $(NRF52_UPLOAD_SN)
+endif
+
+%.upload: $(OUT_HEX) %.$(TARGET)
```
Comment (g-oikonomou) [inline arch/cpu/nrf52840/Makefile.nrf52840:103]: I wonder if some of those virtual targets used for programming devices can be made `.PHONY`

**Comment by toshiba-bril** [comment]
Hi Atis,

TSCH and CSMA have been tested on a 12-node internal testbed and are operational. 
We don't have any public testbed results or reliability metrics yet. We have observed very intermittent TSCH desyncing when adaptive timesync is on, however we haven't looked into this yet and are unsure if it is a real problem. We are planning on testing this more extensively in the near future.

There were no plans to support BLE when this port was created, as this would require a dependency on the SoftDevice, while we specifically wanted to avoid that.
If the SoftDevice was added to the build some device drivers would need to be modified to use their API to access the hardware peripherals, and the 802.15.4 radio driver would not support coexistence as it does not use their timeslot API, the SD blocks direct access to the radio otherwise.

We would greatly appreciate anyone trying out this port and reporting any radio-related issues they might find.

File: arch/cpu/nrf52840/Makefile.nrf52840 (hunk 14 above)
Comment (toshiba-bril) [reply arch/cpu/nrf52840/Makefile.nrf52840:108]: Added `nrf-upload-sequence` as `.PHONY`

File: arch/cpu/nrf52840/Makefile.nrf52840 [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+
+.PHONY: nrf-upload-sequence
+
+# Upload via J-Link using nrfjprog 
+
+ifdef NRF52_UPLOAD_SN
+  NRFJPROG_FLAGS += --snr $(NRF52_UPLOAD_SN)
+endif
+
+%.upload: $(OUT_HEX) %.$(TARGET)
```
Comment (g-oikonomou) [inline arch/cpu/nrf52840/Makefile.nrf52840:108]: We can lose the dep on `%.$(TARGET)` here too, no?

File: arch/cpu/nrf52840/Makefile.nrf52840 [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+NRF52_SDK_INC_PATHS += components/libraries/fifo
+NRF52_SDK_INC_PATHS += components/libraries/log
+NRF52_SDK_INC_PATHS += components/libraries/log/src
+NRF52_SDK_INC_PATHS += components/drivers_nrf/nrf_soc_nosd
+
+EXTERNALDIRS += $(addprefix $(NRF5_SDK_ROOT)/, $(NRF52_SDK_INC_PATHS))
+
+include $(ARCH_PATH)/cpu/arm/cortex-m/cm4/Makefile.cm4
+
+.PHONY: nrf-upload-sequence
```
Comment (g-oikonomou) [inline arch/cpu/nrf52840/Makefile.nrf52840:100]: What about the rest? `%.upload`, `%.upload-all`, `nrf-upload-sequence`?

File: .gitmodules [hunk 17]
```diff
@@ -28,3 +28,6 @@
 [submodule "arch/cpu/cc26xx-cc13xx/lib/cc2640r2-sdk"]
 	path = arch/cpu/cc26x0-cc13x0/lib/cc2640r2-sdk
 	url = https://github.com/contiki-ng/cc2640r2-sdk.git
+[submodule "arch/cpu/nrf52840/lib/nrf52-sdk"]
+	path = arch/cpu/nrf52840/lib/nrf52-sdk
+	url = https://github.com/alexstanoev/nrf-sdk-contiki-ng.git
```
Comment (g-oikonomou) [inline .gitmodules:33]: All good from my side. We'll only need to relocate this before merging

**Pull request merged**